### PR TITLE
Add L2 rollback

### PR DESCRIPTION
**Pull request by BenjaminLudwigSAP**
There is currently no revert logic for when SelfIP or subnet route creation/deletion fails.
This can lead to a problem when a SelfIP fails to be created (e.g. due to conflicting pool member) but the subnet route it replaces has already been deleted. In that case the deleted subnet route won't be restored.

In order to enable reverting subnet route creation/deletion, the subnet routes that existed before the L2 flow runs must be known beforehand, so that the tasks know what the desired state is. For this to work, two things are done:
- Fetching the previously existing subnet routes and SelfIPs via a dedicated flow
- Running one task for every single subnet route creation/deletion (instead of running one task to create all at once)

These changes are made in the first commit of this PR.
The second commit adds the actual revert methods.

Please read the commit messages for more detailed explanations.

### CODE REVIEW

File: octavia_f5/controller/worker/flows/f5_flows.py [hunk 1]
```diff
@@ -23,98 +23,201 @@
...
+        :param subnets_that_need_routes: Subnets for which subnet routes must exist
         """
+        host = store['bigip'].hostname
+        network = store['network'].id
+        preexisting_selfips = store['existing_selfips']
+        preexisting_subnet_routes = store['existing_subnet_routes']
+
+        # find SelfIPs that are expected but not existing
+        selfips_to_create = [port for port in needed_selfips if port.id not in [p.id for p in preexisting_selfips]]
+        LOG.debug(f"{host}: SelfIPs to add for network {network.id}: {[p.id for p in selfips_to_create]}")
```
Comment (notandy) [inline octavia_f5/controller/worker/flows/f5_flows.py]: `id` of `network` is dereferenced twice, once in L172, making `network` a string and then again in L178, causing

```
ERROR octavia_f5.controller.worker.l2_sync_manager [-] Failed running sync_l2_selfips_and_subnet_routes_flow for host ...: 'str' object has no attribute 'id'
```

File: octavia_f5/controller/worker/flows/f5_flows.py [hunk 2]
```diff
@@ -23,98 +23,201 @@
...
+
+        # make SelfIP creation subflow
+        ensure_selfips_subflow = unordered_flow.Flow('ensure-selfips-subflow')
+        for selfip in selfips_to_create:
+            ensure_selfip_task = f5_tasks.EnsureSelfIP(name=f"ensure-selfip-{selfip.id}", inject={'port': selfip})
+            ensure_selfips_subflow.add(ensure_selfip_task)
+
+        # find subnet routes for subnets that need them but don't have any yet
+        subnet_route_network_part = f5_tasks.get_subnet_route_name(network.id, '')
```
Comment (notandy) [inline octavia_f5/controller/worker/flows/f5_flows.py:191]: `network` is a string

File: octavia_f5/controller/worker/flows/f5_flows.py [hunk 3]
```diff
@@ -23,98 +23,201 @@
...
+            r['name'][len(subnet_route_network_part):] for r in preexisting_subnet_routes
+            if r['name'].startswith(subnet_route_network_part)
+        ]
+        subnets_to_create_routes_for = [s for s in subnets_that_need_routes
+                                        if s not in subnets_of_preexisting_subnet_routes]
+
+        # make subnet route creation subflow
+        ensure_subnet_routes_subflow = unordered_flow.Flow('ensure-subnet-routes-subflow')
+        for subnet_id in subnets_to_create_routes_for:
+            subnet_route_name = f5_tasks.get_subnet_route_name(network.id, subnet_id)
```
Comment (notandy) [inline octavia_f5/controller/worker/flows/f5_flows.py:204]: `network` is a string

File: octavia_f5/controller/worker/flows/f5_flows.py [hunk 4]
```diff
@@ -23,98 +23,201 @@
...
+
+        # find subnet routes for subnets that need them but don't have any yet
+        subnet_route_network_part = f5_tasks.get_subnet_route_name(network.id, '')
+        subnets_of_preexisting_subnet_routes = [
+            r['name'][len(subnet_route_network_part):] for r in preexisting_subnet_routes
+            if r['name'].startswith(subnet_route_network_part)
+        ]
+        subnets_to_create_routes_for = [s for s in subnets_that_need_routes
+                                        if s not in subnets_of_preexisting_subnet_routes]
+
```
Comment (notandy) [inline octavia_f5/controller/worker/flows/f5_flows.py:200]: `make_remove_selfips_and_subnet_routes_flow` has a log for subnet routes to be removed(L138), I would expect that create do the same for subnet routes to be added.

File: octavia_f5/controller/worker/tasks/f5_tasks.py [hunk 5]
```diff
@@ -356,58 +399,81 @@ def execute(self, network: f5_network_models.Network,
...
-                delete_all=False):
-
-        subnets_that_need_routes = []
-        if not delete_all:
-            subnets_that_need_routes = [subnet for subnet in network.subnets if not subnet_in_selfips(subnet, selfips)]
-
-        # prefix of subnet routes that belong to this network
-        subnet_route_network_part = get_subnet_route_name(network.id, '')
+                subnet_route, existing_subnet_routes):
+        route_name_name = subnet_route['name']
```
Comment (notandy) [inline octavia_f5/controller/worker/tasks/f5_tasks.py]: `subnet_route` is a string

**Review by notandy — CHANGES_REQUESTED** [review]
Hi, so I've found multiple exceptions and stopped debugging at some point.

**Comment by notandy** [comment]
Since this is a very large PR and it's hard to debug every corner case, I'd like to request following addition:

Add a unit tests for all (modified/added) Tasks and Flows, that
* Test Creation of Subnet routes
* Test deletion of subnet routes
* Test creation of selfips
* test removal of selfips
* test replacement of selfip with route (e.g. remove a VIP from subnet)
* test replacement of route with selfip (e.g. added a VIP to subnet)

**Comment by m-kratochvil** [comment]
From my end this looks good, it's a lot of code which I can't reliably review but I at least went through and checked for anything wrong like typos or mistakes in the comments but everything looks good. We had few discussions with Benjamin on Slack about the networking concepts related to routes and self-IPs on Big-IP so I trust everything is implemented correctly. But I will let Andy or Vadim press the "approve" button, from my end this PR is approved.

**Comment by velp** [comment]
Looks good to me. Pretty much everything is covered with unit tests (checked coverage report). The logic looks OK for me too. Let's wait final word from Andy.

**Review by velp — APPROVED** [review]

File: octavia_f5/controller/worker/tasks/f5_tasks.py [hunk 6]
```diff
@@ -206,7 +204,7 @@ def execute(self, network: f5_network_models.Network,
             res.raise_for_status()
             return res.json()
 
-        # Update if dict differs from on-device state
+        # Otherwise update existing subnet route (if our route isn't a subset)
```
Comment (notandy) [inline octavia_f5/controller/worker/tasks/f5_tasks.py]: the comment seems c&p from the subnet task? (should refer self-ips instead)

**Review by notandy — APPROVED** [review]
I tested a bit, and apart from minor comment typos it looks good.

**Comment by m-kratochvil** [comment]
Checked the last three commits, look good to me, I don't see anything wrong.
